chromadb/rate_limiting/admission_controller.py

```diff
@@ -0,0 +1,358 @@
+
+    def resize_waiters(self) -> None:
+        self.check_invariants()
+        if (self.tail + 1) % len(self.waiters) != self.head:
+            raise RuntimeError(
+                "Admission controller invariants violated: resize when not full"
+            )
+
+        new_size = len(self.waiters) * 2
+        new_waiters: list[Optional[Waiter]] = [None] * new_size
+        new_idx = 0
+        for i in range(len(self.waiters)):
+            idx = (self.head + i) % len(self.waiters)
+            if idx == self.tail:
+                break
+            if self.waiters[idx] is not None:
+                w: Waiter = cast(Waiter, self.waiters[idx])
+                new_waiters[new_idx] = w
+                self.waiters[idx] = None
+                w.index = new_idx
+                new_idx += 1
+
+        self.waiters = new_waiters
+        self.head = 0
+        self.tail = new_idx
+        self.check_invariants()
+
+    def adjust_queue_mode(self, now: float) -> None:
+        if self.head == self.tail:
+            self.last_empty = now
+            self.queue_mode = self.FIFO_MODE
+        elif self.queue_mode == self.FIFO_MODE and self.last_empty + self.N < now:
+            self.queue_mode = self.LIFO_MODE
+
+    def strip_head_nils(self) -> None:
+        while self.head != self.tail and self.waiters[self.head] is None:
+            self.head = (self.head + 1) % len(self.waiters)
+
+    def strip_tail_nils(self) -> None:
+        while self.head != self.tail:
+            idx = (self.tail - 1) % len(self.waiters)
+            if self.waiters[idx] is not None:
+                break
+            self.tail = idx
+
+    def check_invariants(self) -> None:
+        for idx, w in enumerate(self.waiters):
+            if w is None:
+                continue
+            if w.index != idx:
+                raise RuntimeError(
+                    "Admission controller invariants violated: waiter at wrong index"
+                )
+
+        if self.head != self.tail:
+            if self.waiters[self.head] is None:
+                raise RuntimeError(
+                    "Admission controller invariants violated: head is nil"
+                )
+            size = len(self.waiters)
+            tail_idx = (self.tail + size - 1) % size
+            if self.waiters[tail_idx] is None:
+                raise RuntimeError(
+                    "Admission controller invariants violated: tail is nil"
+                )
+
+
+global_admitted = 0
+global_denied = 0
+lock = threading.Lock()
+ac = AdmissionControllerImpl(parallelism=10)  # Number of workers
+
+NUM_REQUESTS = 100
+
+
+def increment_admitted() -> None:
+    global global_admitted
+    with lock:
+        global_admitted += 1
+
+
+def increment_denied() -> None:
+    global global_denied
+    with lock:
+        global_denied += 1
+
+
+def stream_requests(n: int) -> tuple[Optional[Ticket], int]:
+    delay = 0.001  # Seconds per request: 1/N to calculate throughput
```

Comment:
I don't believe this comment is accurate.  We'll have 1/51ms of throughput, not 1/1ms.

Reply:
How did you arrive here? I was thinking if each request is 1 ms, we'll have 1 request/ms or 1000 requests/sec

Reply:
Because your request service time is 50ms on line 323

Reply:
I was thinking about the throughput at which new requests will get added to the queue